presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoinsUtils.java

```diff
@@ -0,0 +1,183 @@
+    {
+        ImmutableSet.Builder<Set<Integer>> sets = ImmutableSet.builder();
+        if (index < totalNodes && existingSet.size() < totalNodes - 1) {
+            ImmutableSet.Builder<Integer> setWithAddedNode = ImmutableSet.builder();
+            setWithAddedNode.addAll(existingSet);
+            setWithAddedNode.add(index);
+            sets.addAll(generatePartitions(setWithAddedNode.build(), totalNodes, index + 1));
+            sets.addAll(generatePartitions(existingSet, totalNodes, index + 1));
+        }
+        else {
+            sets.add(existingSet);
+        }
+        return sets.build();
+    }
+
+    public static JoinNode createJoinAccordingToPartitioning(JoinGraphNode joinGraph, Set<Integer> leftSources, PlanNodeIdAllocator idAllocator)
+    {
+        List<PlanNode> sources = joinGraph.getSources();
+        ImmutableList.Builder<PlanNode> leftNodes = ImmutableList.builder();
+        List<Symbol> leftSymbols = new ArrayList<>();
+        ImmutableList.Builder<PlanNode> rightNodes = ImmutableList.builder();
+        List<Symbol> rightSymbols = new ArrayList<>();
+        for (int i = 0; i < sources.size(); i++) {
+            PlanNode source = sources.get(i);
+            if (leftSources.contains(i)) {
+                leftNodes.add(source);
+                leftSymbols.addAll(source.getOutputSymbols());
+            }
+            else {
+                rightNodes.add(source);
+                rightSymbols.addAll(source.getOutputSymbols());
+            }
+        }
+
+        ImmutableList.Builder<JoinNode.EquiJoinClause> leftCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> rightCriteria = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> spanningCriteria = ImmutableList.builder();
+        for (JoinNode.EquiJoinClause equiJoinClause : joinGraph.getCriteria()) {
+            Set<Symbol> equiJoinClauseSymbols = ImmutableSet.of(equiJoinClause.getLeft(), equiJoinClause.getRight());
+            if (leftSymbols.containsAll(equiJoinClauseSymbols)) {
+                leftCriteria.add(equiJoinClause);
+            }
+            else if (rightSymbols.containsAll(equiJoinClauseSymbols)) {
+                rightCriteria.add(equiJoinClause);
+            }
+            else {
+                spanningCriteria.add(equiJoinClause);
+            }
+        }
+
+        ImmutableList.Builder<Expression> leftFilters = ImmutableList.builder();
+        ImmutableList.Builder<Expression> rightFilters = ImmutableList.builder();
+        ImmutableList.Builder<Expression> spanningFilters = ImmutableList.builder();
+        for (Expression filter : joinGraph.getFilters()) {
+            Set<Symbol> filterSymbols = SymbolReferenceExtractor.getSymbolReferences(filter).stream()
+                    .map(Symbol::from)
+                    .collect(toImmutableSet());
+            if (leftSymbols.containsAll(filterSymbols)) {
```

Comment:
this probably won't work for lambdas, e.g:
```
left symbols: a, b
filterExpression: apply(a, x -> x)
then
filterSymbols: a, x
and
!leftSymbols.containsAll(filterSymbols)
```
See https://github.com/Teradata/presto/pull/547/commits/ec4269b0e174783491a8e9a936f7ef6cb9640635#r114991216. You might need to fix `DependencyExtractor.SymbolBuilderVisitor`